### PR TITLE
[go_router_builder] Adds name parameter to TypedGoRoute and supports its generation.

### PR DESCRIPTION
**Pull request by dancamdev**
This PR allows for generating `TypedGoRoutes` by defining a custom name and forwarding it to the underlying `GoRoute` this is needed when working with Analytics services such as Google analytics that use the name to log the page in their systems.

Fixes [#120102](https://github.com/flutter/flutter/issues/120102)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by dancamdev** [comment]
I'd need a hand here, CI is failing because `go_router_builder` is relying on the upstream `go_router` version, @chunhtai would it be ok to rely on the local go_router instead? What is the recommended way to do this otherwise?

**Comment by chunhtai** [comment]
You need to separate out two pr, one for go_router and one for go_router_builder. The go_router one need to be merged first before the ci in go_router_builder can pass

**Review by chunhtai — CHANGES_REQUESTED** [review]
please separate out two prs

**Comment by dancamdev** [comment]
Needs [#3702](https://github.com/flutter/packages/pull/3702) to be merged before tests can pass and this pr can be merged.

**Comment by chunhtai** [comment]
Hi @dancamdev, what's the status of this pr?

File: packages/go_router_builder/test/test_inputs/_go_router_builder_test_input.dart [hunk 1]
```diff
@@ -325,3 +325,28 @@ class IterableDefaultValueRoute extends GoRouteData {
   IterableDefaultValueRoute({this.param = const <int>[0]});
   final Iterable<int> param;
 }
+
+@ShouldGenerate(r'''
+RouteBase get $namedRoute => GoRouteData.$route(
+      path: '/named-route',
+      name: 'namedRoute',
```
Comment (chunhtai) [inline packages/go_router_builder/test/test_inputs/_go_router_builder_test_input.dart:332]: Can you also add a test for with name that requires dart string escape?

**Review by chunhtai — COMMENTED** [review]
This branch seems old, can you rebased off latest main?

File: packages/go_router_builder/test/test_inputs/_go_router_builder_test_input.dart (hunk 1 above)
Comment (dancamdev) [reply packages/go_router_builder/test/test_inputs/_go_router_builder_test_input.dart:332]: Good point, sure!

**Review by hangyujin — APPROVED** [review]
LGTM

**Pull request merged**